sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/WalkedTypePath.scala

```diff
@@ -0,0 +1,50 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.sql.catalyst
+
+case class WalkedTypePath(walkedPaths: Seq[String] = Nil) extends Serializable {
```

Comment:
shall we use a mutable list for better performance?

Reply:
It was to address diverged paths for map key and value, but we can also copy instance via cloning internal list if necessary. Will address.

Reply:
Addressed via 90df8a3 - I found it a bit complicated to maintain the list without polluting, so please take a look at the change and let me know if you would like to roll back to immutable one if performance gain doesn't seem to have more value than complexity.